providers/common/exchange/dh_exch.c

```diff
@@ -42,22 +42,26 @@ static int dh_init(void *vpdhctx, void *vdh)
 {
     PROV_DH_CTX *pdhctx = (PROV_DH_CTX *)vpdhctx;
 
+    if (pdhctx == NULL)
+        return 0;
     DH_free(pdhctx->dh);
     pdhctx->dh = vdh;
-    DH_up_ref(pdhctx->dh);
-
-    return pdhctx->dh != NULL;
+    if (vdh == NULL)
+        return 0;
+    return DH_up_ref(pdhctx->dh);
```

Comment:
Please do the DH_up_ref before the pdhctx is modified.

Reply:
Apologies, misunderstood the intent.  Fixed now.